File: game24.cpp
#include <iostream>
#include <string>
#include <stack>
#include <vector>
#include <map>
#include <set>
#include <algorithm>
#include <thread>
#include <future>
#include <cstdlib>

int NUM_THREADS = 1;

enum class ItemType {
    Numeric,
    Operator_ADD,
    Operator_SUB,
    Operator_MUL,
    Operator_DIV,
};

class Item {
public:
    Item(char o): op_(o) {
        switch (o) {
            case '+': tp_ = ItemType::Operator_ADD; break;
            case '-': tp_ = ItemType::Operator_SUB; break;
            case '*': tp_ = ItemType::Operator_MUL; break;
            default: tp_ = ItemType::Operator_DIV; break;
        }
    }

    Item(int n): number_(n), tp_(ItemType::Numeric) {}

    std::string to_string() const {
        if(tp_==ItemType::Numeric) return std::to_string(number_);
        return std::string(1, op_);
    }

    int number() const {
        return number_;
    }

    bool is_number() const {
        return tp_ == ItemType::Numeric;
    }

    ItemType type() const {
        return tp_;
    }

    bool try_operator(double a, double b, double& out) const {
        if(tp_==ItemType::Numeric) {
            return false;
        }

        if(tp_==ItemType::Operator_ADD) {
            out = a + b;
            return true;
        }

        if(tp_==ItemType::Operator_SUB) {
            out = a - b;
            return true;
        }

        if(tp_==ItemType::Operator_MUL) {
            out = a * b;
            return true;
        }

        // DIV
        if(b==0) return false;

        out = a / b;
        return true;
    }

    bool operator < (const Item& rhs) const {
        return number_ < rhs.number_;
    }

    bool operator >= (const Item& rhs) const {
        return !(*this < rhs);
    }

private:
    int number_ = 0;
    char op_;
    ItemType tp_;
};


//void output(const std::vector<Item*>& xx) {
//    std::cout << "OUTPUT" << std::endl;
//    for(auto* x: xx) {
//        std::cout << x->to_string() << " ";
//    }
//    std::cout << std::endl;
//
//}


class SuffixFormula {
public:
    bool push(Item* item) {
        if(!invalid_) return false;

        sequence_.push_back(item);

        if(item->is_number()) {
            stack_.push(static_cast<double>(item->number()));
            return true;
        }

        if(stack_.size() < 2) {
            invalid_ = false;
            return false;
        }

        auto n1 = stack_.top();
        stack_.pop();
        auto n2 = stack_.top();
        stack_.pop();

        double n3 = 0;

        invalid_ = item->try_operator(n2, n1, n3);
        if(invalid_) {
            stack_.push(n3);
            op_amount_++;
        }

        return invalid_;
    }

    bool invalid() const {
        return invalid_;
    }

    double value() const {
        return stack_.top();
    }

    void to_normalize(std::vector<Item*>& result) {
        for(Item* item: sequence_) {
            result.push_back(item);
        }

        normalize_exchange(result);
        normalize_same_op(result, ItemType::Operator_ADD);
        normalize_same_op(result, ItemType::Operator_MUL);
        normalize_exchange_special(result, ItemType::Operator_SUB);
        normalize_exchange_special(result, ItemType::Operator_DIV);
    }

    std::string to_infix() {
        std::vector<Item*> sequence;
        to_normalize(sequence);

        std::vector<std::string> results;
        for(auto* item: sequence) {
            results.push_back(item->to_string());
        }

        while (results.size()>1) {
            for(size_t index=0; index<results.size(); index++) {
                if(results[index] == "+" || results[index] == "-" || results[index] == "*" || results[index] == "/") {
                    std::string res;
                    res += "(";
                    res += results[index-2];
                    res += " ";
                    res += results[index];
                    res += " ";
                    res += results[index-1];
                    res += ")";

                    results.erase(results.begin()+index-2);
                    results.erase(results.begin()+index-2);
                    results.erase(results.begin()+index-2);

                    results.insert(results.begin()+index-2, res);
                    break;
                }
            }
        }

        return results[0].substr(1, results[0].length()-2);
    }

private:
    std::stack<double> stack_;
    std::vector<Item*> sequence_;
    size_t op_amount_ = 0;
    bool invalid_ = true;

    size_t find_sub_sequence_index(const std::vector<Item*>& sequence, size_t pos) {
        int num_amount = 0;
        int op_amount = 0;

        while(pos>=0) {
            if(sequence[pos]->is_number()) num_amount++;
            else op_amount++;

            if(num_amount == op_amount+1) return pos;
            pos--;
        }

        return 0;
    }

    void normalize_exchange(std::vector<Item*>& sequence) {
        size_t length = sequence.size();
        for(size_t index=2; index<=length-1; index++) {
            if (sequence[index]->type() == ItemType::Operator_ADD || sequence[index]->type() == ItemType::Operator_MUL) {
                Item* item1 = sequence[index-2];
                Item* item2 = sequence[index-1];

                if(!item2->is_number()) {
                    auto sub_index_2 = find_sub_sequence_index(sequence, index-1);
                    if(sequence[sub_index_2-1]->is_number()) continue;

                    auto sub_index_1 = find_sub_sequence_index(sequence, sub_index_2-1);
                    if(*sequence[sub_index_2] >= *sequence[sub_index_1]) continue;

                    std::vector<Item*> sub_seq_2;
                    auto sub_seq_2_length = index-1-sub_index_2+1;
                    sub_seq_2.reserve(sub_seq_2_length);
                    auto iter = sequence.begin();
                    for(size_t i=0; i<sub_seq_2_length; i++) {
                        sub_seq_2.push_back(sequence[sub_index_2]);
                        sequence.erase(iter+sub_index_2);
                    }

                    sequence.insert(sequence.begin()+sub_index_1, sub_seq_2.begin(), sub_seq_2.end());
                    continue;
                }

                if(item1->is_number()) {
                    if(*item2 < * item1) {
                        sequence[index-2] = item2;
                        sequence[index-1] = item1;
                    }

                    continue;
                }

                auto sub_index = find_sub_sequence_index(sequence, index);
                sequence.erase(sequence.begin()+index-1);
                sequence.insert(sequence.begin()+sub_index, item2);
            }
        }
    }

    void normalize_exchange_special(std::vector<Item*>& sequence, ItemType op_tp) {
        size_t length = sequence.size();
        for(size_t index=0; index<length; index++) {
            if(index+4 > length-1) return;

            auto a = sequence[index];
            auto b = sequence[index+1];
            auto c = sequence[index+2];
            auto d = sequence[index+3];
            auto e = sequence[index+4];

            if(a->is_number() && b->is_number() && c->type() == op_tp && d->is_number() && e->type() == op_tp) {
                if(*d < *b) {
                    sequence[index+1] = d;
                    sequence[index+3] = b;
                }
            }
        }
    }

    void normalize_same_op(std::vector<Item*>& sequence, ItemType op_tp) {
        std::vector<size_t> op_indexes;
        std::vector<Item*> op_items;

        for(size_t i=0; i<sequence.size(); i++) {
            if(sequence[i]->type() == op_tp) {
                op_indexes.push_back(i);
                op_items.push_back(sequence[i]);
            }
        }

        if(op_indexes.size()!=op_amount_) return;

        for(size_t i=0; i<op_amount_; i++) {
            sequence.erase(sequence.begin() + op_indexes[i] - i);
        }

        auto cmp = [](Item* a, Item* b) {return *a < *b;};
        std::sort(sequence.begin(), sequence.end(), cmp);
        sequence.insert(sequence.end(), op_items.begin(), op_items.end());
    }
};


class Operators {
public:
    Operators(int amount): amount_(amount) {
        for(int i=0; i<amount; i++) {
            levels_[i] = 0;
        }
    }

    ~Operators() {
        for(auto* item: operators_) delete item;
    }

    std::vector<Item*> get() {
        std::vector<Item*> opts;
        if(levels_[0] > 3) return opts;

        for(int i=0; i<amount_; i++) {
            opts.push_back(operators_[levels_[i]]);
        }

        levels_[amount_-1] += 1;

        for(int i=amount_-1; i>0; i--) {
            if(levels_[i] > 3) {
                levels_[i] = 0;
                levels_[i-1] += 1;
            }
        }

        return opts;
    }

private:
    int amount_;
    const std::vector<Item*> operators_{new Item('+'), new Item('-'), new Item('*'), new Item('/')};
    std::map<int, int> levels_;
};

bool is_invalid_sequence(const std::vector<Item*>& seq) {
    int number_amount = 0;
    int op_amount = 0;

    for(auto* item: seq) {
        if(item->is_number()) number_amount++;
        else op_amount++;

        if(op_amount >= number_amount) return false;
    }

    return true;
}

void find(int target, std::vector<Item*> numbers, std::vector<std::vector<Item*>> ops, std::promise<std::set<std::string>>& p) {
    std::set<std::string> results;
    std::cout << "thread: " << std::this_thread::get_id() << ", ops len, " << ops.size() << std::endl;
    for(auto& op: ops) {
        std::vector<Item*> items;
        items.insert(items.end(), numbers.begin(), numbers.end());
        items.insert(items.end(), op.begin(), op.end());

//        auto cmd = [](Item* a, Item* b) {return *a < *b;};

        std::sort(items.begin(), items.end());
        do {
            if(!is_invalid_sequence(items)) continue;

            SuffixFormula sf;
            for(Item* item: items) {
                if(!sf.push(item)) break;
            }

            if(!sf.invalid()) continue;

            double x = sf.value() - target;
            if(x >= -1e-3 && x <= 1e-3) {
                results.insert(sf.to_infix());
            }

        } while (std::next_permutation(items.begin(), items.end()));

    }
    p.set_value(results);

}

int main(int argc, char* argv[]) {
    int target, amount, number;
    std::vector<Item*> numbers;

    target = std::atoi(argv[1]);
    amount = argc - 2;

    for(int i=1; i<=amount; i++) {
        number = std::atoi(argv[1+i]);
        numbers.push_back(new Item(number));
    }

    Operators ops(amount-1);
    std::vector<std::vector<Item*>> all_ops;
    while (true) {
        auto op = ops.get();
        if(op.empty()) break;

        all_ops.push_back(op);
    }

    std::set<std::string> results;

    size_t step = all_ops.size() / NUM_THREADS;

    std::vector<std::vector<Item*>> splitted_ops[NUM_THREADS];
    for(int i=0; i<NUM_THREADS; i++) {
        splitted_ops[i].insert(splitted_ops[i].end(), all_ops.begin()+step*i, all_ops.begin()+step*(i+1));
    }

    splitted_ops[NUM_THREADS-1].insert(splitted_ops[NUM_THREADS-1].end(), all_ops.begin()+step*NUM_THREADS, all_ops.end());

    std::promise<std::set<std::string>> p[NUM_THREADS];
    std::future<std::set<std::string>> f[NUM_THREADS];
    std::thread ts[NUM_THREADS];

    for(int i=0; i<NUM_THREADS; i++) {
        f[i] = p[i].get_future();
        ts[i] = std::thread(find, target, numbers, splitted_ops[i], std::ref(p[i]));
    }

    std::cout << "wait thread ..." << std::endl;
    for(int i=0; i<NUM_THREADS; i++) {
        ts[i].join();
        auto xx = f[i].get();

        results.insert(xx.begin(), xx.end());
    }

    if(results.empty()) {
        std::cout << "no results" << std::endl;
    } else {
        for(auto& s: results) {
            std::cout << s << " = " << target << std::endl;
        }
    }

    for(auto* item: numbers) delete item;

    return 0;
}

File: game24.py
# -*- coding: utf-8 -*-

import itertools
from operator import add, sub, mul, div

OPERATIONS = {
    '+': add,
    '-': sub,
    '*': mul,
    '/': div,
}

PRIORITY = {
    '+': ['+', '-'],
    '-': ['+', '-'],
    '*': ['*', '/'],
    '/': ['*', '/'],
}


# 后缀表达式， 运算符在后面
# 比如中缀表达式 1 * ( 2 + 3 ) 对应的后缀表达式是 1 2 3 + *
# 后缀表达式的优点就是 不用借助括号， 表达式求值的顺序是确定的
# SuffixFormula 类用来表示一个后缀表达式
# 对其求值，转成中缀表达式，最重要的就是变换表达式，除去重复


class SuffixFormula(object):
    def __init__(self):
        # 每次压入一个元素，遇到 运算符 立即求值。
        # 这里最适合用stack数据结构，但用list模拟
        self.stack = []
        # sequence 仅仅是保存了原始序列
        self.sequence = []
        self.op_amount = 0
        # push进来的元素是否能组成一个合法的表达式
        self.invalid = True

    def push(self, item):
        # 把item压入stack, item 要么是 数字， 要么是操作符
        # 如果是操作符，那么立即将栈顶的两个元素出栈，
        # 再把根据此运算符计算出的结果，压栈
        # 返回bool值， False表示此次入栈的item 会导致这是一个非法的表达式

        """

        :rtype: bool
        """
        self.sequence.append(item)

        if not self.invalid:
            return False

        if item not in OPERATIONS:
            self.stack.append(item)
            return True

        # 遇到操作符，栈顶两元素出栈，并且将计算结果再入栈
        # 这里会遇到的错误情况： 前面不够两个元素，或者0作了除数
        try:
            n1 = self.stack.pop(-1)
            n2 = self.stack.pop(-1)
        except IndexError:
            self.invalid = False
            return False

        try:
            n3 = OPERATIONS[item](float(n2), n1)
        except ZeroDivisionError:
            self.invalid = False
            return False

        self.stack.append(n3)
        self.op_amount += 1
        return True

    def value(self):
        # 获取表达式的值，对于合法的表达式，在push的过程中已经完成了求值过程
        # 就是stack中唯一的元素
        """

        :rtype: int
        """
        return self.stack[0]

    @staticmethod
    def _find_sub_sequence_index(sequence, pos):
        # 找子表达式的起始位置， pos是此子表达式的 运算符
        # 这里要利用到表达式的特性： 一个合法的表达式序列中，数字的数量=运算符数量+1
        # 比如 12+3+4+
        # 第一个+，一个运算符，需要往前面找两个数字，刚好是 1 2
        # 第二个+，也需要往前面找两个数字，但找到一个数字后，又遇到一个运算符，
        # 此时还需要找 2+1-1 个数字

        if sequence[pos] not in OPERATIONS:
            raise ValueError("item at post is not operator")

        num_amount = 0
        op_amount = 0
        index = pos
        while index >= 0:
            if sequence[index] in OPERATIONS:
                op_amount += 1
            else:
                num_amount += 1

            if num_amount == op_amount + 1:
                return index

            index -= 1

        raise ValueError("can not find sub index")

    # 下面 _normalize 开头的方法是对 表达式序列 按照一定规则进行变换，从而达到去重的目的
    def _normalize_remove_brackets(self, sequence, op):
        # 去掉多余括号
        # 有这几个表达式：
        # 1. 1 + (2 + (3 - 4))  =>  1 2 3 4 - + +
        # 2. (1 + 2) + (3 - 4)  =>  1 2 + 3 4 - +
        # 3. 1 + 2 + 3 - 4      =>  1 2 + 3 + 4 -
        # 因为括号，导致后缀表达式不一样， 但中缀表达式实质上是一样的
        # 这样的情况会对下面的 交换 处理有干扰，
        # 所以去掉括号，统一处理成3 那种情况

        loop = True
        while loop:
            loop = False
            for index in range(3, len(sequence)):
                if sequence[index] != op:
                    continue

                # 前面的是 同等优先级的 运算符，才能这么处理
                if sequence[index - 1] not in PRIORITY[op]:
                    continue

                if sequence[index - 2] in OPERATIONS:
                    to_index = self._find_sub_sequence_index(sequence, index - 2)
                else:
                    to_index = index - 2

                sequence.pop(index)
                sequence.insert(to_index, op)
                loop = True
                # 因为后面的运算符放到前面去，可能导致前面的也可以去掉括号
                # 所以这里需要一遍一遍的遍历
                break

        return sequence

    def _normalize_exchange(self, sequence):
        # 对 '+', '*' 的 两边 进行交换
        # 有这几个表达式：
        # 1. 4 + 3 * (1 + 2)  =>  4 3 1 2 + * +
        # 2. 4 + 3 * (2 + 1)  =>  4 3 2 1 + * +
        # 3. 4 + (1 + 2) * 3  =>  4 1 2 + 3 * +
        # 4. 3 * (1 + 2) + 4  =>  3 1 2 + * 4 +
        # 虽然后缀表达式不相同，但中缀表达式确实重复
        # 这里就要变换后缀表达式，从而使得这三种写法的中缀表达式，都变换成一种。
        # 如何定位 两边的元素 及 交换规则
        # AB+, +前面是两个数字， 那么两边就是A 和 B，把小的放前面
        # ABC+D*, * 前面是数字和+, 那么 两边就是 D 和 +的子表达式，把数字D放在子表达式前面

        for index in range(2, len(sequence)):
            if sequence[index] == '+' or sequence[index] == '*':
                # changeable
                item1 = sequence[index - 2]
                item2 = sequence[index - 1]

                if item2 in OPERATIONS:
                    # item2 是运算符，对应这种情况 ? + *
                    sub_index_2 = self._find_sub_sequence_index(sequence, index - 1)
                    if sequence[sub_index_2 - 1] not in OPERATIONS:
                        # 对应这种情况 1 2 3 + *, 数字已经在子表达式前面了，不用交换
                        continue

                    # 对应这种情况 1 2 + 3 4 + *, 两个子表达式按照第一个数字大小决定是否需要交换
                    sub_index_1 = self._find_sub_sequence_index(sequence, sub_index_2 - 1)
                    if sequence[sub_index_2] >= sequence[sub_index_1]:
                        continue

                    # 把sub_index_2开始的子表达式放到前面去
                    sub_seq_2 = []
                    for _ in range(index - 1 - sub_index_2 + 1):
                        sub_seq_2.insert(0, sequence.pop(sub_index_2))

                    for s in sub_seq_2:
                        sequence.insert(sub_index_1, s)

                    continue

                if item1 not in OPERATIONS:
                    # item2, item1 都是数字，对应这种情况 1 2 +,把小的放前面
                    if item1 > item2:
                        sequence[index - 2], sequence[index - 1] = sequence[index - 1], sequence[index - 2]

                    continue

                # item2 是数字，item1 是运算符，对应这种情况 * 1 +
                # 这时候要找到*子表达式的起始位置，然后把1放过去
                sub_index = self._find_sub_sequence_index(sequence, index)
                sequence.pop(index - 1)
                sequence.insert(sub_index, item2)

        return sequence

    def _normalize_exchange_special(self, sequence, op):
        # 上面只是对 '+', '*' 等做的处理，
        # '-', '/' 的处理比较复杂，而且去掉括号后，还要改变其他运算符，本来就不是重复的，所以可以不处理
        # 但有一种特殊情况，需要处理
        # 1 - 2 - 3  =>  1 2 - 3 -
        # 1 - 3 - 2  =>  1 3 - 2 -
        # 这两种情况其实是一样，算作重复。 对于除法也一样。
        # 只对后面两个数字 按照大小排序处理

        for index in range(len(sequence)):
            try:
                a = sequence[index]
                b = sequence[index + 1]
                c = sequence[index + 2]
                d = sequence[index + 3]
                e = sequence[index + 4]
            except IndexError:
                break

            if a not in OPERATIONS and b not in OPERATIONS and c == op and d not in OPERATIONS and e == op:
                if d < b:
                    sequence[index+1] = d
                    sequence[index+3] = b

        return sequence

    def _normalize_same_op(self, sequence, op):
        # 相同的运算符，只有 '+' 或者 '*'
        # 1 + (2 + (3 + 4))  =>  1 2 3 4 + + +
        # (1 + 2) + (3 + 4)  =>  1 2 + 3 4 + +
        # 1 + 2 + 3 + 4      =>  1 2 + 3 + 4 +
        # 这几种情况是一样的，但上面的 exchange 还是把 后缀表达式处理成不同的。
        # 因此，这里直接简单的处理成第一种情况， 数字按照大小排序

        op_amount = 0
        for s in sequence:
            if s == op:
                op_amount += 1

        if op_amount != self.op_amount:
            return sequence

        for _ in range(op_amount):
            sequence.remove(op)

        sequence.sort()
        sequence.extend([op] * op_amount)
        return sequence

    def to_normalize(self):
        sequence = self.sequence[:]

        sequence = self._normalize_exchange(sequence)
        # sequence = self._normalize_remove_brackets(sequence, '+')
        # sequence = self._normalize_remove_brackets(sequence, '*')
        # sequence = self._normalize_exchange(sequence)
        sequence = self._normalize_same_op(sequence, '+')
        sequence = self._normalize_same_op(sequence, '*')
        sequence = self._normalize_exchange_special(sequence, '-')
        sequence = self._normalize_exchange_special(sequence, '/')

        return sequence

    def to_infix(self):
        # 返回 中缀表达式 字符串
        sequence = self.to_normalize()

        while True:
            if len(sequence) == 1:
                break

            for index in range(len(sequence)):
                if sequence[index] in OPERATIONS:
                    express = '({0} {1} {2})'.format(
                        sequence[index - 2], sequence[index], sequence[index - 1]
                    )

                    sequence.pop(index - 2)
                    sequence.pop(index - 2)
                    sequence.pop(index - 2)

                    sequence.insert(index - 2, express)
                    break

        return sequence[0][1:-1]


# 公式生成器，暴力枚举所有组合
class FormulaGenerator(object):
    def __init__(self, *numbers):
        self.numbers = numbers

    @staticmethod
    def _is_invalid_sequence(seq):
        number_amount = 0
        op_amount = 0
        for s in seq:
            if s in OPERATIONS:
                op_amount += 1
            else:
                number_amount += 1

            if op_amount >= number_amount:
                return False

        return True

    def __iter__(self):
        ops = itertools.product(['+', '-', '*', '/'], repeat=len(self.numbers) - 1)
        for op in ops:
            sequence = itertools.permutations(self.numbers + op)
            for seq in sequence:
                if self._is_invalid_sequence(seq):
                    yield seq


def find(target, *numbers):
    results = set()

    fg = FormulaGenerator(*numbers)
    for seq in fg:
        sf = SuffixFormula()

        for s in seq:
            if not sf.push(s):
                break

        # 上面把生成的公式序列放入 SuffixFormula 后
        # 如果这个序列不合法，就尝试下一个序列
        if not sf.invalid:
            continue

        # 序列是合法的公式，那么其值和 target 接近，就认为相等
        # 因为有除法，会有小数
        if abs(sf.value() - target) > 1e-3:
            continue

        # 如果是相同的公式，就会在这里过滤掉
        results.add(sf.to_infix())

    return list(results)
